[IR] Add `llvm.sincos` intrinsic

MacDue opened the pull request:
This adds the `llvm.sincos` intrinsic, legalization, and lowering.

The `llvm.sincos` intrinsic takes a floating-point value and returns both the sine and cosine (as a struct).

```
declare { float, float }          @llvm.sincos.f32(float  %Val)
declare { double, double }        @llvm.sincos.f64(double %Val)
declare { x86_fp80, x86_fp80 }    @llvm.sincos.f80(x86_fp80  %Val)
declare { fp128, fp128 }          @llvm.sincos.f128(fp128 %Val)
declare { ppc_fp128, ppc_fp128 }  @llvm.sincos.ppcf128(ppc_fp128  %Val)
declare { <4 x float>, <4 x float> } @llvm.sincos.v4f32(<4 x float>  %Val)
```

The lowering is built on top of the existing FSINCOS ISD node, with additional type legalization to allow for f16, f128, and vector values.

MacDue committed:
[IR] Add `llvm.sincos` intrinsic

This adds the `llvm.sincos` intrinsic, legalization, and lowering.

The `llvm.sincos` intrinsic takes a floating-point value and returns
both the sine and cosine (as a struct).

```
declare { float, float }          @llvm.sincos.f32(float  %Val)
declare { double, double }        @llvm.sincos.f64(double %Val)
declare { x86_fp80, x86_fp80 }    @llvm.sincos.f80(x86_fp80  %Val)
declare { fp128, fp128 }          @llvm.sincos.f128(fp128 %Val)
declare { ppc_fp128, ppc_fp128 }  @llvm.sincos.ppcf128(ppc_fp128  %Val)
declare { <4 x float>, <4 x float> } @llvm.sincos.v4f32(<4 x float>  %Val)
```

The lowering is built on top of the existing FSINCOS ISD node, with
additional type legalization to allow for f16, f128, and vector values.

MacDue commented:
Note: This is part of https://discourse.llvm.org/t/rfc-add-struct-returning-intrinsics-for-math-functions-with-output-pointers/81164

MacDue commented:
> ```diff
> R = PromoteFloatRes_FSINCOS(N);
> ```

Not sure why, but my clang format (`18.1.8`) disagrees here (so I had to manually fix this).

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -5577,6 +5577,16 @@ void SelectionDAGLegalize::PromoteNode(SDNode *Node) {
+    for (unsigned ResNum = 0; ResNum < Node->getNumValues(); ResNum++)
+      Results.push_back(
+          DAG.getNode(ISD::FP_ROUND, dl, OVT, Tmp2.getValue(ResNum),
+                      DAG.getIntPtrConstant(0, dl, /*isTarget=*/true)));
```
Host this out of loop

tschuett commented on llvm/docs/LangRef.rst:
```diff
@@ -15694,6 +15698,47 @@ trapping or setting ``errno``.
+and :ref:`llvm.cos <t_llvm_cos>` on the argument.
+
+The first result is the sine of the argument and the second result is the cosine
+of the argument.
```
When specified with the fast-math-flag ‘afn’, the result may be approximated using a less accurate calculation.

sin and cos support fast-math-flags, but sincos does not?

arsenm replied:
as a struct output, this is not an FPMathOperator

tschuett replied:
But when I replace sincos by sin + cos then I get fast-math-flags, see IRTranslator changes.

MacDue replied:
I think the definition of `FPMathOperator` would need to be changed for `sincos` (currently it does not allow struct values).

tschuett replied:
`MachineInstr::copyFlagsFromInstruction(CI)` for the sincos intrinsic.

MacDue replied:
You mean special-case `sincos` without updating the definition of `FPMathOperator`?

MacDue replied:
I had a look into this, but it seems fast-math flags are fairly defendant on `FPMathOperator`.  It looks like a while ago it was extended to support ArrayTypes (https://reviews.llvm.org/D69161), so I've created a similar PR to extend it to support homogeneous literal struct types: https://github.com/llvm/llvm-project/pull/110506 (which would allow it for `llvm.sincos`).

MacDue replied:
Added the note about `afn` now that #110506 has landed (though the actual approximation is not implemented in this patch).

tschuett commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -2340,6 +2340,13 @@ bool IRTranslator::translateKnownIntrinsic(const CallInst &CI, Intrinsic::ID ID,
   }
+  case Intrinsic::sincos: {
+    ArrayRef<Register> VRegs = getOrCreateVRegs(CI);
+    MIRBuilder.buildFSincos(VRegs[0], VRegs[1],
```
This is more complicated than needed, see  case Intrinsic::fptosi_sat: below.

tschuett replied:
Ignore please.

tschuett commented:
Please add G_SINCOS to https://llvm.org/docs/GlobalISel/GenericOpcode.html.

RKSimon commented:
We've never made any progress vectorizing intrinsics with aggregate return types like this (e.g. overflow intrinsics)

@fhahn @alexey-bataev Do you think we can easily support this or should we consider alternatives? We've also had problems with handling arguments that take pointers (sincos / modf etc.) so I guess its just an annoying problem in general.......

MacDue commented:
> We've never made any progress vectorizing intrinsics with aggregate return types like this (e.g. overflow intrinsics)
> 
> @fhahn @alexey-bataev Do you think we can easily support this or should we consider alternatives? We've also had problems with handling arguments that take pointers (sincos / modf etc.) so I guess its just an annoying problem in general.......

I have a patch WIP patch to support this, I could share as a draft now.

arsenm commented:
Can this deal with sin/cos pairs, like div/rem are already handled? I assume the vectorizer can deal with making sure they're likely to be used in codegen

MacDue commented:
> > We've never made any progress vectorizing intrinsics with aggregate return types like this (e.g. overflow intrinsics)
> > @fhahn @alexey-bataev Do you think we can easily support this or should we consider alternatives? We've also had problems with handling arguments that take pointers (sincos / modf etc.) so I guess its just an annoying problem in general.......
> 
> I have a patch WIP/experimental patch to support this, I could share as a draft now.

Here's my (experimental/WIP) implementation for vectorizing (homogeneous) structs: https://github.com/llvm/llvm-project/pull/109833. It works for my current set of test cases (based on functions like sincos).

tschuett commented:
We commonly add a test when importing a new intrinsics: https://github.com/llvm/llvm-project/pull/88240.
We have for the first time in a while an intrinsic which returns two values. A test would be nice to have to make sure we got everything right.

MacDue committed:
Fixups

tschuett commented:
Thanks for the test and the documentation update!

MacDue committed:
Add note about afn

MacDue commented:
Kind ping :slightly_smiling_face:

RKSimon commented:
Do you want to get this in before #109833 ? Whichever way you do it, I'd like to see sincos vectorisation test coverage.

MacDue commented:
> Do you want to get this in before #109833 ? Whichever way you do it, I'd like to see sincos vectorisation test coverage.

I've split them so they can go in any order. Just #109833 alone won't vectorize `llvm.sincos`, the vectorizer also needs to be told it's possible (which is just a small patch), so that's when I'm planning on adding sincos vectorization tests.

RKSimon commented on llvm/test/CodeGen/AArch64/llvm.sincos.ll:
```diff
@@ -0,0 +1,516 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 2
+; RUN: llc -mtriple=aarch64-gnu-linux < %s | FileCheck -check-prefixes=CHECK %s
```
Are there aarch64/arm targets that don't have sincos that we can test on as well?

MacDue replied:
I've added tests where the libcall is unavailable (and so the intrinsic is expanded to individual sin/cos calls) :+1:

arsenm commented on llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-sincos.ll:
```diff
@@ -0,0 +1,106 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+; RUN: llc -O0 -mtriple=aarch64-linux-gnu -global-isel -stop-after=irtranslator %s -o - | FileCheck %s
+
```
Test flag preservation

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -6918,12 +6918,23 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
     SDVTList VTs = DAG.getVTList(ValueVTs);
-    setValue(&I,
-             DAG.getNode(ISD::FFREXP, sdl, VTs, getValue(I.getArgOperand(0))));
+    setValue(&I, DAG.getNode(Opcode, sdl, VTs, getValue(I.getArgOperand(0))));
```
Flags lost?

MacDue replied:
Should be forwarded now :+1:

MacDue committed:
Fixups

- Flag propagation
- Legalize to individual sin/cos calls when FSINCOS unavailable
- More tests

arsenm commented on llvm/test/CodeGen/AArch64/llvm.sincos.ll:
```diff
@@ -0,0 +1,936 @@
+  %result = call { <2 x double>, <2 x double> } @llvm.sincos.v2f64(<2 x double> %a)
+  %result.1 = extractvalue { <2 x double>, <2 x double> } %result, 1
+  ret <2 x double> %result.1
+}
```
Should have a dag test with flags, not sure if there's a good way to see a manifested codegen difference (I guess could have a user that depends on an upstream nnan flag)

MacDue replied:
I added a simple test that uses the `-debug=isel` output to check it: https://github.com/llvm/llvm-project/pull/109825/files#diff-0a696752af71ad6ceb144da9fac6f2ac9b1c13ea233d700484b044b9c07fc36d (marked with `REQUIRES: asserts`).

MacDue committed:
Add SDAG flags test

sdesmalen-arm reviewed (commented):
This intrinsic and the reason for having it makes sense to me. The code looks fine too, only left a few minor comments.

MacDue commented on llvm/test/CodeGen/AArch64/llvm.sincos.ll:
```diff
@@ -0,0 +1,936 @@
+  %result.1 = extractvalue { half, half } %result, 1
+  ret half %result.1
+}
+
```
The `<2 x half>` tests already hit the widening code path (it's widened to `<4 x half>`)

MacDue replied:
But only `<3 x float>` hits the unrolling path right now, which needs some fixups :+1:

MacDue commented on llvm/test/CodeGen/AArch64/llvm.sincos.ll:
```diff
@@ -0,0 +1,936 @@
+  ret half %result.0
+}
+
+define half @test_sincos_f16_only_use_cos(half %a) {
```
I think they can probably be removed (I don't think they're testing new code paths, I just initially was following the style of the frexp tests).

MacDue commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4749,6 +4755,14 @@ void DAGTypeLegalizer::WidenVectorResult(SDNode *N, unsigned ResNo) {
     Res = WidenVecRes_Ternary(N);
     break;
+  case ISD::FSINCOS: {
+    if (!unrollExpandedOp())
+      Res = WidenVecRes_FSINCOS(N);
+    for (unsigned ResNum = 0; ResNum < N->getNumValues(); ResNum++)
+      SetWidenedVector(SDValue(N, ResNum), Res.getValue(ResNum));
+    Res = SDValue();
```
Yes, it should unroll of possible (which is does for `<3 x float>`). There are some cases where it does not unroll right now, but probably should, but that's a general limitation of the "should unroll" check.

MacDue replied:
> Is it worth creating a WidenVecRes_UnaryOpWithTwoResults for this? (it seems FREXPR is missing in this switch statement)

It may be, but I think creating something to handle both `FFREXP` and `FSINCOS` may take a little more care. `FSINCOS` is easy as both results are the same type (so have the same legalization action), but that might be something that needs handling differently for `FFREXP` (which returns an int and a float).

MacDue replied:
Done this now, and enabled widening for `FREXPR` (with two AArch64 tests).

MacDue committed:
Fixups

- Test <3 x float> case and fix unrolling
- Fix langref nit
- Remove redundant tests

MacDue committed:
Add and use `WidenVecRes_UnaryOpWithTwoResults` for SINCOS + FREXPR

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4429,6 +4435,22 @@ SDValue DAGTypeLegalizer::SplitVecOp_VECTOR_HISTOGRAM(SDNode *N) {
+  EVT OtherVT = N->getValueType(OtherNo);
+  if (getTypeAction(OtherVT) == TargetLowering::TypeWidenVector) {
+    SetWidenedVector(SDValue(N, OtherNo), SDValue(WidenNode, OtherNo));
+  } else {
```
This needs a check that the other operand is also a vector.

MacDue replied:
Every `_UnaryOpWithTwoResults` method currently assumes both results are vectors (of the same element count), which has to hold for both `FSINCOS` and `FREXPR`. I'm not sure it should be further generalized while there are no nodes with mismatched values. Maybe the name could be changed to make this assumption explicit?

sdesmalen-arm replied:
I guess that's not really by design but rather because there are no operations that return a vector and a non-vector result. Perhaps you can just add an explicit assert that the type of `N`/`WidenNode` is a vector?

MacDue replied:
Added an assert at the top of the function :+1:

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4448,6 +4470,8 @@ void DAGTypeLegalizer::WidenVectorResult(SDNode *N, unsigned ResNo) {
     if (!TLI.isOperationLegalOrCustomOrPromote(N->getOpcode(), WideVecVT) &&
         TLI.isOperationExpand(N->getOpcode(), VT.getScalarType())) {
       Res = DAG.UnrollVectorOp(N, WideVecVT.getVectorNumElements());
+      if (N->getNumValues() == 2)
```
For operations that return > 2 values (which I appreciate you might not be able to test), this doesn't do the right thing now. If you change this condition to `> 1`, then the code in `ReplaceOtherWidenResult` should handle the case the operation has more results. This currently fails the assert, but I think can be implemented with a loop for the other values.

MacDue replied:
I'm not sure this should be implemented now. `DAG.UnrollVectorOp()` also asserts that the number of results is 1 or 2. I could just add an assert after the `if` that the number of results is 1 and leave further extension to when someone needs it?

MacDue replied:
I guess not -- I've generalized this, but yeah, there are no nodes that test this path right now.

MacDue committed:
Fixups

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -774,6 +775,45 @@ SDValue DAGTypeLegalizer::SoftenFloatRes_FFREXP(SDNode *N) {
+  auto [ReturnVal, Chain] = TLI.makeLibCall(DAG, LC, NVT, Ops, CallOptions, DL,
+                                            /*Chain=*/SDValue());
+  unsigned ResNo = 0;
+  for (SDValue OutPtr : {StackSlotSin, StackSlotCos}) {
```
It might be cleaner to do this with a lambda:
```
auto CreateStackLoad = [&](SDValue StackSlot) {
  int FrameIdx = cast<FrameIndexSDNode>(OutPtr)->getIndex();
  auto PtrInfo = MachinePointerInfo::getFixedStack(DAG.getMachineFunction(), FrameIdx);
  return DAG.getLoad(NVT, DL, Chain, OutPtr, PtrInfo);
};
SetSoftenedFloat(SDValue(N, 0), CreateStackLoad(StackSlotSin));
SetSoftenedFloat(SDValue(N, 1), CreateStackLoad(StackSlotCos));
```

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -2899,6 +2943,18 @@ SDValue DAGTypeLegalizer::PromoteFloatRes_FFREXP(SDNode *N) {
+  SDValue Op = GetPromotedFloat(N->getOperand(0));
+  SDValue Res = DAG.getNode(N->getOpcode(), SDLoc(N), {NVT, NVT}, Op);
+
+  for (unsigned ResNum = 0; ResNum < N->getNumValues(); ResNum++)
```
(style) `for (unsigned ResNum = 0, NumValues = N->getNumValues(); ResNum < NumValues; ++ResNum)`

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -3304,6 +3364,26 @@ SDValue DAGTypeLegalizer::SoftPromoteHalfRes_FFREXP(SDNode *N) {
+
+  // Convert back to FP16 as an integer.
+  ISD::NodeType Truncate = GetPromotionOpcode(NVT, OVT);
+  for (unsigned ResNum = 0; ResNum < N->getNumValues(); ResNum++) {
```
(style) `for (unsigned ResNum = 0, NumValues = N->getNumValues(); ResNum < NumValues; ++ResNum)`

RKSimon reviewed (commented):
A few minor comments

MacDue committed:
Style fixups

RKSimon reviewed (approved):
LGTM - cheers
